feat: add `array/float16`

udaykakade25 opened the pull request:
progresses #7347 and #7273 

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds array/float16 folder

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #7273 #7347 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

This is the third PR. First two PRs are interconnected to each other. In order to merge this PR, we need to first merge PR 1: #7273 and #7347 

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

udaykakade25 committed:
feat: add array/float16

udaykakade25 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

udaykakade25 commented:
Hello @Planeshifter and @kgryte sir,
With this PR, i have completed `array/float16`.
Please do review and merge PR along with two previously mentioned PRs.

So, i can start working on adding float16 datatype support for  `array/base`, `array/*`, `strided/8` and `ndarray/*`

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,538 @@
+
+// MAIN //
+
+class Float16Array {
```
This entire implementation should be refactored. We do not use ES6+. All code must be written in ES5. Furthermore, all methods and properties should be documented. See `array/bool` and `array/complex128` for what we expect.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,538 @@
+		return Float16Array.fromFloat16(this._buffer[index]);
+	}
+
+  get buffer() {
```
This file also does not conform to our style conventions and includes mixed tabs and spaces. Please follow our development guide and ensure that your local environment is properly setup. Our automated linting and EditorConfig configuration would have flagged all these issues.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,538 @@
+	const offset = value !== undefined ? Number(value) : 0;
+
+	if (offset < 0 || offset > this.length) {
+		throw new RangeError('Offset is out of bounds');
```
This is not how we format error messages.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,538 @@
+		(ArrayBuffer.isView(source) && typeof source.length === 'number') ||
+		(source instanceof Float16Array))
+	) {
+		const len = Math.min(source.length, this.length - offset);
```
We do not use `Math` built-ins. Everything should be implemented in terms of stdlib functionality.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,538 @@
+	return this._entriesGenerator();
+}
+
+*_entriesGenerator() {
```
`entries` should be implemented using iterators directly. Use of generators is not allowed in ES5.

kgryte reviewed (changes requested):
@udaykakade25 Thank you for working on this. I left a few initial comments. The TL;DR is that the entire implementation needs to be reworked to be implemented in ES5. I strongly encourage you to emulate the style and approach used in `array/bool` and other custom array types implemented in stdlib.

udaykakade25 committed:
Refactored Polyfill.js

Signed-off-by: Uday Kakade <141299403+udaykakade25@users.noreply.github.com>

udaykakade25 commented:
> @udaykakade25 Thank you for working on this. I left a few initial comments. The TL;DR is that the entire implementation needs to be reworked to be implemented in ES5. I strongly encourage you to emulate the style and approach used in `array/bool` and other custom array types implemented in stdlib.

Hello @kgryte sir, I have refactored the whole polyfill.js file that follows stdlib style. Thanks for reviewing!!

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,903 @@
+* @throws {TypeError} Must provide valid input
+*/
+function Float16Array( input, byteOffset, length ) {
+    var i;
```
This file should be all TAB indentation. Currently, you are using space indentation.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,903 @@
+    if ( typeof input === 'number' ) {
+        this._buffer = new Uint16Array( input );
+    } else if ( input instanceof ArrayBuffer ) {
+        offset = byteOffset >>> 0;
```
This isn't correct. `>>> 0` casts `byteOffset` to `uint32`, but typed arrays can have more than `2**32 - 1` elements.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,903 @@
+* @readonly
+* @type {ArrayBuffer}
+*/
+Object.defineProperty( Float16Array.prototype, 'buffer', {
```
As in `array/bool`, you should be using our utilities for defining properties. E.g., `utils/define-nonenumerable-read-only-accessor`. Right now, `buffer` is enumerable, which it shouldn't be.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,903 @@
+*/
+Float16Array.from = function( source, mapFn, thisArg ) {
+    if ( source == null ) {
+        throw new TypeError( 'invalid argument. Source cannot be null or undefined. Value: `' + source + '`.' );
```
For error formatting, as in `array/bool`, you should be using `@stdlib/string/format`.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,903 @@
+    var result = new Float16Array( this.length );
+    var i;
+    var val;
+    if ( typeof callback !== 'function' ) {
```
You need to be using stdlib assertion utilities for validation here and everywhere else. E.g., `assert/is-function`.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,903 @@
+* @param {*} [thisArg] - Context
+* @returns {boolean} Whether all pass
+*/
+Float16Array.prototype.every = function( callback, thisArg ) {
```
Please define methods in alphabetical order. This makes searching for a particular easier for future readers of this code.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,903 @@
+*
+* @returns {Object} Iterator
+*/
+Float16Array.prototype.keys = function() {
```
We don't use anonymous functions. We used named functions to assist in making stack traces more legible.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,903 @@
+
+    this.length = this._buffer.length;
+}
+
```
You are missing static constructor properties, such the constructor name and `BYTES_PER_ELEMENT`.

kgryte commented:
Also, for this PR to move forward, CI should be passing. Currently, CI is failing on linting, benchmarks, examples, and tests.

udaykakade25 committed:
Refactored polyfill.js

Signed-off-by: Uday Kakade <141299403+udaykakade25@users.noreply.github.com>

udaykakade25 commented:
> Also, for this PR to move forward, CI should be passing. Currently, CI is failing on linting, benchmarks, examples, and tests.

Hello @kgryte sir,
I have changes that follow the standards below
TABS Indentation ✅
Alphabetical Order ✅
Offset >>> 0 fix ✅
Readonly added ✅
Error formatting ✅
Utilities Added ✅
No Anonymous functions ✅
Missing Statistics Added ✅

Regarding failed CI checks:
Just Linting errors in Readme.md and Repl.txt are remaining to fix.
Rest other issues are encountered because the `assert\has-float16array-support` #7347 and `assert/is-float16array` #7273  PRs is still pending.

udaykakade25 committed:
Update package.json

Signed-off-by: Uday Kakade <141299403+udaykakade25@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/polyfill.js:
```diff
@@ -0,0 +1,1044 @@
+		this.set( index, val );
+	};
+}
+Float16Array.prototype.fill = function( value, start, end ) {
```
Use `setReadOnly` here.

udaykakade25 committed:
Update polyfill.js

Signed-off-by: Uday Kakade <141299403+udaykakade25@users.noreply.github.com>

gururaj1512 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into arrayFloat16Final

gururaj1512 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into arrayFloat16Final

gururaj1512 committed:
fix: benchmarks

gururaj1512 committed:
refactor: implementation

gururaj1512 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into arrayFloat16Final

gururaj1512 committed:
refactor: implmentation and add test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2501 @@
+		end = len;
+	}
+	for ( i = start; i < end; i++ ) {
+		buf[ i ] = toWord( value );
```
```suggestion
	v = toWord( value );
	for ( i = start; i < end; i++ ) {
		buf[ i ] = v;
```

kgryte replied:
Need to declare `v` above.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2501 @@
+		}
+	} else {
+		fromIndex = 0;
+	}
+	buf = this._buffer;
+	for ( i = fromIndex; i < this._length; i++ ) {
+		if ( searchElement === fromWord( buf[ i ] ) ) {
+			return true;
+		}
+	}
```
```suggestion
	buf = this._buffer;
	v = toWord( searchElement );
	for ( i = fromIndex; i < this._length; i++ ) {
		if ( v === buf[ i ] ) {
			return true;
		}
	}
```

kgryte replied:
Need to declare `v` above.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2501 @@
+		fromIndex = 0;
+	}
+	buf = this._buffer;
+	for ( i = fromIndex; i < this._length; i++ ) {
```
Same comments as `includes` above.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2501 @@
+	out = [];
+	for ( i = 0; i < this._length; i++ ) {
+		out.push( fromWord( buf[i] ) );
+	}
```
```suggestion
	out = new Float64Array( this._length );
	for ( i = 0; i < this._length; i++ ) {
		out[ i ] = fromWord( buf[ i ] );
	}
```

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2501 @@
+		fromIndex = this._length - 1;
+	}
+	buf = this._buffer;
+	for ( i = fromIndex; i >= 0; i-- ) {
```
Same comment as `includes` above.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2501 @@
+	}
+	buf = this._buffer;
+	if ( arguments.length === 0 ) {
+		buf.sort();
```
This won't work as you think it does. You are sorting Uint16 integers; however, negative values will be "greater" than non-negative values, as the sign bit is set in negative values. Hence, ascending/descending will not work as expected. You actually need to use an internal comparator here which compares `float16` values directly.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2501 @@
+	for ( i = 0; i < this._length; i++ ) {
+		out.push( fromWord( buf[ i ] ).toLocaleString( loc, opts ) );
+	}
+	return out.join( ',' );
```
```suggestion
	out = new Float64Array( this._length );
	for ( i = 0; i < this._length; i++ ) {
		out[ i ] = fromWord( buf[ i ] );
	}
	return out.toLocaleString( loc, opts );
```

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2501 @@
+	*/
+	function compare( a, b ) {
+		return compareFcn( a, b );
+	}
```
```suggestion
	out = new this.constructor( this );
	if ( arguments.length === 0 ) {
		return out.sort();
	}
	return out.sort( compare );
```

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2501 @@
+	for ( i = 0; i < this._length; i++ ) {
+		out.push( fromWord( buf[i] ) );
+	}
+	return out.join( ',' );
```
Same comment as above. Leverage `Float64Array` here.

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2501 @@
+	}
+	out = new this.constructor( out );
+	buf = out._buffer; // eslint-disable-line no-underscore-dangle
+	buf[ index ] = toWord( value );
```
```suggestion
	out = new this.constructor( this );
	out.set( value, index );
```

gururaj1512 committed:
chore: apply suggested changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
feat: enhance Float16Array constructor to handle Float16Array instances

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: add readme

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: update README and examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
I see that doctesting is failing due to missing `Float16Array` support. Will take a look.

gururaj1512 committed:
chore: add and refactor benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 commented:
> I see that doctesting is failing due to missing `Float16Array` support. Will take a look.

Should I look into that?
Also if `Float16Array` doctesting support added lot of cleanup can be carried out by removing `.get()` from examples.

gururaj1512 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

kgryte commented:
@gururaj1512 Chicken and the egg problem here. Looked into doctest support. Essentially we first need to merge this PR, and then we can add doctest support. And once doctest support is added, we can circle back and clean up the examples.

gururaj1512 commented:
@kgryte, That makes sense. This PR still need some cleanup. I'll mark it up for review once cleaned-up

kgryte commented:
@gururaj1512 I've made some changes locally. Would you like me to go ahead and push those up before continuing further?

gururaj1512 commented:
@kgryte, I'm fine with it. You can push your changes.

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
@gururaj1512 Done.

gururaj1512 committed:
refactor: use format for benchmark names

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -183,11 +183,7 @@ function Float16Array() {
-				buf = new Uint16Array( arg._buffer ); // eslint-disable-line no-underscore-dangle
-			} else {
-				buf = fromArray( new Uint16Array( arg.length ), arg );
-			}
```
@kgryte,
```js
arr = new Float16Array( new Float16Array( [1.05] ) );
console.log(arr.get(0));
// returns 1.0498046875
```
Ideally it should work as above but currently this return `NaN`.
So I have added this statement.
Removing this results in erreneous behaviour.

kgryte replied:
I think the issue is that the builtin Float16Array won't have a private buffer property.

gururaj1512 replied:
```js
	if ( isFloat16Array( arg ) ) {
		buf = new Uint16Array( arg.buffer, arg.byteOffset, arg.length );
	} else {
		buf = fromArray( new Uint16Array( arg.length ), arg );
	}
```
I think this should work well then?

kgryte replied:
Yes, that should work!

gururaj1512 committed:
fix: handle Float16Array input in Float16Array constructor

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: improve buffer initialization in Float16Array constructor

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
refactor: fix docs and types

gururaj1512 committed:
fix: clarify return type

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/main.js:
```diff
@@ -0,0 +1,2512 @@
+			buf = new Uint16Array( arg );
+		} else if ( isCollection( arg ) ) {
+			if ( isFloat16Array( arg ) ) {
+				buf = new Uint16Array( arg.buffer, arg.byteOffset, arg.length );
```
```suggestion
				buf = copy( new Uint16Array( arg.buffer, arg.byteOffset, arg.length ), new Uint16Array( arg.length ) );
```

kgryte replied:
We need to ensure that we are not sharing memory with the input array.

kgryte replied:
There is the `blas/base/gcopy` routine:

```js
buf = gcopy( arg.length, view, 1, 0, out, 1, 0 );
```

gururaj1512 committed:
feat: add copyArray function to copy elements between Uint16Arrays

kgryte commented on lib/node_modules/@stdlib/array/float16/lib/copy_array.js:
Why add a custom function here rather than use `blas/base/gcopy`?

gururaj1512 replied:
Updated to use `blas/base/gcopy` 👍🏻

gururaj1512 committed:
fix: remove copyArray function and replace with gcopy in Float16Array constructor

gururaj1512 committed:
fix: lint error

udaykakade25 commented:
Hey @kgryte and @gururaj1512,
Can you refactor and merge this `array/float16` PR?

Started working on adding support for `array/*`,  `array/base/*`, `strided/*`, and `ndarray/*`.
Taking this as ref: https://github.com/stdlib-js/stdlib/wiki/GSoC-2024-%E2%80%90-Jaysukh-Makvana

Would love to show my recent progress in the next office hour meet :)

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into arrayFloat16Final

kgryte commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

kgryte commented:
/stdlib help

kgryte committed:
bench: fix allocations to ensure 1:1 comparison across array types

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: fix allocations to ensure 1:1 comparison across array types

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: fix allocations to ensure 1:1 comparison across array types

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: use string interpolation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: fix examples and remove `get` documentation

kgryte committed:
fix: ensure `set` is consistent with actual `Float16Array`

kgryte commented:
Unfortunately, there were a number of things wrong with this PR, especially in the understanding of how the polyfill is intended to work. While `array/bool` and `array/complex64` were intended to be examples of how to author the package, those packages are also purely userland array types; however, `Float16Array` is a built-in type which we are polyfilling, which means that we need to match its behavior exactly. We cannot simply make the polyfill an accessor array, while the builtin is not. This is thus requiring a lot of refactoring, which I am still doing.

So, while I would have liked to get this PR in, I'll likely need more time to do so.

kgryte committed:
Merge branch 'develop' into arrayFloat16Final

kgryte committed:
chore: clean-up tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: fix examples

kgryte committed:
test: remove obsolete tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: update condition

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: update condition

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: update assertion messages

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
test: fix failing tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: relax assertions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: fix assertion

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: fix assertion

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: fix assertion

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
refactor: update to support exporting built-in and move polyfill tests and benchmarks to subfolders

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
style: re-enable lint rule

kgryte reviewed (commented):
Okay. I think this PR is now ready provided CI passes for everything but the doctest return annotations, which will be resolved once this PR is merged.

I will check back later today, and, if everything looks good, I'll go ahead and merge.

kgryte committed:
style: resolve lint failures

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented:
Checked. All CI failures are expected due to missing doctest support, which I will add in a follow-up commit. As such, I'll go ahead and merge this PR. Thanks, @udaykakade25 and @gururaj1512, for helping get this over the finish line!
